Improve documentation of `@validate_call()` (#10455)
pydantic/pydantic: docs/concepts/validation_decorator.md

can be enabled by using a [custom configuration](#custom-configuration).

Type coercion like this can be extremely helpful, but also confusing or not desired.
See [Coercion and strictness](#coercion-and-strictness) for a discussion of `@validate_call`'s limitations
in this regard.
!!! note "Validating the return value"
    By default, the return value of the function is **not** validated. To do so, the `validate_return` argument
    of the decorator can be set to `True`.

## Function signatures

The `@validate_call` decorator is designed to work with functions using all possible parameter configurations and
all possible combinations of these:

* Positional or keyword arguments with or without defaults.
* Variable positional arguments defined via `*` (often `*args`).
* Variable keyword arguments defined via `**` (often `**kwargs`).
* Keyword-only arguments: arguments after `*,`.
* Positional-only arguments: arguments before `, /` (new in Python 3.8).
The [`validate_call()`][pydantic.validate_call] decorator is designed to work with functions
using all possible [parameter configurations][parameter] and all possible combinations of these:

> To demonstrate all the above parameter types:

Author: I'm inclined to remove this long example which imo doesn't really add any value, as such examples can be found in the linked Python docs above (https://docs.python.org/3/glossary.html#term-parameter).

Reviewer: Honestly, let's leave it. I think it's sometimes helpful to do a bit of Python review to help folks who may be unfamiliar with such syntax, but realize it's relevant for their use case. I think it also makes people happy to see that we support such a wide variety of cases here.

That being said, I do think it'd be great if we could link to the relevant Python docs.

Reviewer: In the API docs I would agree, but I think more in depth examples on the concepts page can be helpful 👍

    """
```

## Parameter types

> Parameter types are inferred from type annotations on the function, or as [`Any`][typing.Any] if unannotated . All types listed in [types](types.md) can be validated, including Pydantic models and [custom types](types.md#custom-types).

Reviewer: ```suggestion
Parameter types are inferred from type annotations on the function, or as [`Any`][typing.Any] if not annotated. All types listed in [types](types.md) can be validated, including Pydantic models and [custom types](types.md#custom-types).
```

Author: Yeah I was unsure about this one. It is used in some places in the Python docs, but  I agree _not annotated_ seems better

As with the rest of Pydantic, types are by default coerced by the decorator before they're passed to the actual function:

```py
from datetime import date

from pydantic import validate_call

@validate_call
def greater_than(d1: date, d2: date, *, include_equal=False) -> date:  # (1)!
    if include_equal:
        return d1 >= d2
    else:
        return d1 > d2

d1 = '2000-01-01'  # (2)!
d2 = date(2001, 1, 1)
greater_than(d1, d2, include_equal=True)
```

1. Because `include_equal` has no type annotation, it will be inferred as [`Any`][typing.Any].

> 2. Although `d1` is a string, it will be converted as a [`date`][datetime.date] object.

Reviewer: ```suggestion
2. Although `d1` is a string, it will be converted to a [`date`][datetime.date] object.
```

Reviewer: Maybe "by the decorator" as well like was in the previous example?

Author: Imo this is a bit redundant, as we already state "... by default coerced by the decorator..." just before the example

Reviewer: Sounds good, but I think we should still do the `as` -> `to`.

Type coercion like this can be extremely helpful, but also confusing or not desired. [Strict mode](strict_mode.md)
can be enabled by using a [custom configuration](#custom-configuration).

!!! note "Validating the return value"
    By default, the return value of the function is **not** validated. To do so, the `validate_return` argument
    of the decorator can be set to `True`.

## Function signatures

The [`validate_call()`][pydantic.validate_call] decorator is designed to work with functions
using all possible [parameter configurations][parameter] and all possible combinations of these:

* Positional or keyword parameters with or without defaults.
* Keyword-only parameters: parameters after `*,`.
* Positional-only parameters: parameters before `, /`.
* Variable positional parameters defined via `*` (often `*args`).
* Variable keyword parameters defined via `**` (often `**kwargs`).

!!! note "[`Unpack`][typing.Unpack] for keyword parameters"
    [`Unpack`][typing.Unpack] and typed dictionaries can be used to annotate the variable

>     keyword parameters of a function:

Reviewer: Maybe this even deserves its own section?

Could you also add some docs for `typing.Unpack` in the API section of the docs?
See https://github.com/pydantic/pydantic/pull/10416#issuecomment-2364225636

Author: > Maybe this even deserves its own section?

unpacked typed dicts are pretty niche, so I don't really know if it is necessary? And I feel it fits nicely with the last parameter type being described:

```
* Variable keyword parameters defined via `**` (often `**kwargs`).
```

Reviewer: > unpacked typed dicts are pretty niche, so I don't really know if it is necessary? And I feel it fits nicely with the last parameter type being described:

That's fine, but I do think we should add some more docs for `typing.Unpack` pre v2.10. I can do that, given that you worked on the PR with the functionality.

Author: Regarding the addition in API docs, it could be added in the stdlib types but I don't think it really makes sense:

`Unpack` is not a type per se, it is a special form having a different meaning depending on the context. It was originally introduced for typevar tuples which we don't support as of today.

Unless your thinking about another section of the API docs?

how_many(number=42)
```

## Accessing the original function

> ## Usage with mypy

Author: Section moved below as it refers to `raw_function`, which wasn't described yet.

# requires: `conn.execute()` that will return `'testing@example.com'`
```

## Compatibility with type checkers

As the [`validate_call()`][pydantic.validate_call] decorator preserves the decorated function's signature,
it should be compatible with type checkers. However, due to current limitations in the Python type system,
the [`raw_function`](#accessing-the-original-function) or other attributes won't be recognized and you will

> need to suppress the error.

Reviewer: Might be helpful to use some of the original copy here - lots of folks are familiar with `mypy` specifically, and we want to specifically reference how to suppress the error:

```
The `validate_call` decorator should work "out of the box" with [mypy](http://mypy-lang.org/) since it's
defined to return a function with the same signature as the function it decorates. The only limitation is that
since we trick mypy into thinking the function returned by the decorator is the same as the function being
decorated; access to the [raw function](#raw-function) or other attributes will require `type: ignore`.
```

### Validation exception

Currently upon validation failure, a standard Pydantic [`ValidationError`][pydantic_core.ValidationError] is raised.
See [model error handling](models.md#error-handling) for details.

This is helpful since its `str()` method provides useful details of the error which occurred and methods like
`.errors()` and `.json()` can be useful when exposing the errors to end users. However, `ValidationError` inherits
from `ValueError` **not** `TypeError`, which may be unexpected since Python would raise a `TypeError` upon invalid
or missing arguments. This may be addressed in future by either allowing a custom error or raising a different
exception by default, or both.

### Coercion and strictness

Pydantic currently leans on the side of trying to coerce types rather than raise an error if a type is wrong,

> see [model data conversion](models.md#data-conversion) and `@validate_call` is no different.

Author: Removed as this was mentioned in the _#Parameter types_ section, and type coercion is not really specific to validate_call, it's a general Pydantic pattern.

Reviewer: Maybe we could link to the data conversion section in the docs above?

I think we should probably leave the validation exception section for folks who might not be super familiar with Pydantic - I think this is helpful context to have.

Author: Added a ref to [model data conversion](models.md#data-conversion).

The validation exception wasn't removed but simplified

### Validation exception

Currently upon validation failure, a standard Pydantic [`ValidationError`][pydantic_core.ValidationError] is raised
(see [model error handling](models.md#error-handling) for details). This is also true for missing required arguments,
where Python normally raises a [`TypeError`][].

### Performance

We've made a big effort to make Pydantic as performant as possible. While the inspection of the decorated
function is only performed once, there will still be a performance impact when making calls to the function
compared to using the original function.

> In many situations, this will have little or no noticeable effect, however be aware that
> [`validate_call()`][pydantic.validate_call] is not an equivalent or alternative to function
> definitions in strongly typed languages; it never will be.

Reviewer: ```suggestion
In many situations, this will have little or no noticeable effect. However, be aware that
[`validate_call()`][pydantic.validate_call] is not an equivalent or alternative to function
definitions in strongly typed languages, and it never will be.
```
